state/execution.go

```diff
@@ -417,15 +416,44 @@ func (blockExec *BlockExecutor) Commit(
 	)
 
 	// Update mempool.
+	ch <- mempoolUpdate{block: block, state: state.Copy(), abciResponse: abciResponse, err: err}
+
+	return res.RetainHeight, nil
+}
+
+// updates the mempool with the latest state asynchronously.
+// it begins flushing the mempool immediately, and then updates the mempool
+// after the new block state is ready.
+func (blockExec *BlockExecutor) asyncUpdateMempool(ch chan mempoolUpdate) {
+	blockExec.mempool.Lock()
+	defer blockExec.mempool.Unlock()
+
+	// while mempool is Locked, flush to ensure all async requests have completed
+	// in the ABCI app before Commit.
```

Comment:
We are supposed to do that _before_ calling ABCI Commit.

Reply:
Your right, thank you!